KEP-2436 Leader Migration: to beta

jiahuif opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This PR graduates https://github.com/kubernetes/enhancements/issues/2436 , Leader Migration for Controller Managers, from alpha to beta, by making following changes:

- Change the feature flag from "alpha, disabled by default" to "beta, enabled by default"
- Copy the `v1alpha` resource definitions to `v1beta`, and add corresponding registration & tests.

#### Special notes for your reviewer:


Leader Migration can graduate to beta because:

- Both cloud-provider-gcp and cloud-provider-aws have tested this feature [1]
- Missing feature has been implemented during 1.22 cycle https://github.com/kubernetes/enhancements/pull/2731 , https://github.com/kubernetes/kubernetes/pull/102711 .

For `vendor/` approver: This PR adds a new module `k8s.io/controller-manager/config/v1beta1` as a result of API graduation. The module needs to be vendor'ed as a part of `k8s.io/controller-manager`.

For feature approver: This PR adds a feature gate defined in `staging/src/k8s.io/controller-manager/pkg/features/kube_features.go` to `pkg/features/kube_features.go` so that `pkg/features/kube_features.go` contains all feature gates of `k8s.io/controller-manager`.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Leader Migration for controller managers graduated to beta.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:

- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/2436
```
[1] [SIG Cloud Provider Meeting Agenda](https://docs.google.com/document/d/1OZE-ub-v6B8y-GuaWejL-vU_f9jsjBbrim4LtTfxssw/), June 23

jiahuif committed:
promote ControllerManagerLeaderMigration to beta.

jiahuif committed:
create LeaderMigrationConfiguration v1beta1

jiahuif commented:
/kind feature

jiahuif commented:
/assign @jpbetz

jpbetz commented:
Do we need conversions or defautls for v1beta1 like exist for v1alpha1?

jiahuif commented:
Conversions from/to v1beta1 are generated. Defaults are applied to internal version only.

jpbetz commented:
/lgtm

cheftako commented on staging/src/k8s.io/controller-manager/pkg/features/kube_features.go:
```diff
@@ -49,5 +50,5 @@ func SetupCurrentKubernetesSpecificFeatureGates(featuregates featuregate.Mutable
 var cloudPublicFeatureGates = map[featuregate.Feature]featuregate.FeatureSpec{
 	IPv6DualStack:                    {Default: true, PreRelease: featuregate.Beta},
-	ControllerManagerLeaderMigration: {Default: false, PreRelease: featuregate.Alpha},
+	ControllerManagerLeaderMigration: {Default: true, PreRelease: featuregate.Beta},
```
I believe this should also exist in pkg/features/kube_features.go. (ala IPv6DualStack)

jiahuif replied:
We have this discussion when working on alpha. The conclusion was that `ControllerManagerLeaderMigration` is a local feature flag to `staging/controller-manger` thus does not need a place in `pkg/features/kube_features.go`. I do not see a reason to inverse such conclusion during alpha-to-beta graduation. Could you illiterate a bit more why we need it in `pkg/features/kube_features.go` now?

cheftako replied:
I don't believe we did have a discussion on this in alpha. You did in fact make this assertion during alpha (https://github.com/kubernetes/kubernetes/pull/99417#issuecomment-789958998) however I did not agree then and still do not. However as an alpha feature I was unwilling to block based on that alone. For beta I do consider this to be blocking. pkg/feature/kube_features is the master list and if its missing from there it will be missing from places like https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/.

jiahuif replied:
Thanks for pointing out the doc. Updating now.

cheftako commented on staging/src/k8s.io/controller-manager/config/v1beta1/types.go:
```diff
@@ -0,0 +1,51 @@
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
```
Are we getting rid of GenericControllerManagerConfiguration (https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/controller-manager/config/v1alpha1/types.go#L25)? If not then I think it should be moving to v1beta1 as well.

jiahuif replied:
`GenericControllerManagerConfiguration` is extracted from common part of corresponding structures of KCM and CCM during the "make staging/controller-manager easier to consume task by @cici37". It is still needed unless we want to rollback any changes during the task. However, `GenericControllerManagerConfiguration` should not go to `v1beta` because:

- The stability is unknown, but it is definitely less than beta
- We are keeping `v1alpha` package following convention of how we graduate resources with feature, it would not hurt to keep GenericControllerManagerConfiguration there.
-  If we are to graduate GenericControllerManagerConfiguration to beta, it would be far out of scope of this PR. We can do this in a separate PR to avoid scope creeping.

cheftako replied:
What does it mean to say LeaderMigrationConfiguration has gone to beta, when its not a K8s resource and is only used from GenericControllerManagerConfiguration which if we don;t move it, is obviously only going to use the v1alpha1 version of LeaderMigrationConfiguration?

jiahuif commented:
/assign @cheftako

jiahuif commented:
/retest

jiahuif commented:
/cc @andrewsykim

jiahuif commented:
Fixing conflicts.

andrewsykim commented on pkg/features/kube_features.go:
```diff
@@ -773,6 +773,13 @@ const (
+	// beta:  v1.22
+	//
+	// Enables Leader Migration for kube-controller-manager and cloud-controller-manager
+	ControllerManagerLeaderMigration featuregate.Feature = "ControllerManagerLeaderMigration"
```
Probably deserves a comment here that this is linked/copied from the one in controller-manager.

Or the controller-manager one should have a comment depending on which one we consider the source of truth.

jiahuif replied:
Updating now.

jiahuif replied:
Note: ` pkg/features/kube_features.go ` is always the source of truth for feature gates.

andrewsykim commented on staging/src/k8s.io/controller-manager/config/v1beta1/types.go:
```diff
@@ -0,0 +1,51 @@
+
+// LeaderMigrationConfiguration provides versioned configuration for all migrating leader locks.
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type LeaderMigrationConfiguration struct {
```
Does any consumers of `LeaderMigrationConfiguration` need to be updated to use the new v1beta1 API now?

andrewsykim replied:
Or I guess not because we only deal with internal types inside our components?

jiahuif replied:
The original `v1alpha` version is kept as of this time. I will update the doc on k/website after this PR is merged.

jiahuif replied:
Short answer: no. Because for now the only usage of this group would be parsing a yaml file, the conversion is taken care for free.

cheftako replied:
GenericControllerManagerConfiguration was the only other type I could find using it (See above). Its one of the reasons I was curious about it getting updated.

jiahuif commented:
/retest

caesarxuchao commented:
/triage accepted

jiahuif commented:
/retest

cheftako commented:
/lgtm
/approve

jiahuif commented:
Fixing feature conflicts, 2nd time.

andrewsykim reviewed (commented):
/approve

for `pkg/features`

jpbetz commented:
/lgtm

BenTheElder commented:
/approve
for vendor/modules.txt

cheftako commented:
/milestone v1.22

jiahuif commented:
/retest

jiahuif committed:
add ControllerManagerLeaderMigration as beta

This feature was in staging/src/k8s.io/controller-manager/pkg/features/kube_features.go but missing here.

jiahuif commented:
fixing conflicts for the 3rd time.

jiahuif commented:
/retest

jpbetz commented:
/lgtm
